sampler.go

```diff
@@ -174,6 +181,177 @@ func (s *rateLimitingSampler) Equal(other Sampler) bool {
+}
+
+// -----------------------
+
+// sampler for the operation level
+type operationSampler struct {
+	probabilisticSampler Sampler
+	lowerBoundSampler    Sampler
+	operation            string
+}
+
+// NewOperationSampler operationSampler is a delegating sampler that applies both probabilisticSampler and
+// rateLimitingSampler. The probabilisticSampler is given higher priority when tags are emitted,
+// ie. if IsSampled() for both samplers return true, the tags for probabilisticSampler will be used.
+func NewOperationSampler(operation string, maxTracesPerSecond, samplingRate float64) (Sampler, error) {
+	probabilisticSampler, err := NewProbabilisticSampler(samplingRate)
+	if err != nil {
+		return nil, err
+	}
+	lowerBoundSampler, err := NewLowerBoundSampler(maxTracesPerSecond, samplingRate)
+	if err != nil {
+		return nil, err
+	}
+	return &operationSampler{
+		probabilisticSampler: probabilisticSampler,
+		lowerBoundSampler:    lowerBoundSampler,
+		operation:            operation,
+	}, nil
+}
+
+func (s *operationSampler) IsSampled(id uint64, operation string) (bool, []Tag) {
+	if sampled, tags := s.probabilisticSampler.IsSampled(id, operation); sampled {
+		return true, tags
+	}
+	return s.lowerBoundSampler.IsSampled(id, operation)
+}
+
+func (s *operationSampler) Close() {
+	s.probabilisticSampler.Close()
+	s.lowerBoundSampler.Close()
+}
+
+func (s *operationSampler) Equal(other Sampler) bool {
+	if o, ok := other.(*operationSampler); ok {
+		return s.operation == o.operation && s.probabilisticSampler.Equal(o.probabilisticSampler) &&
+			s.lowerBoundSampler.Equal(o.lowerBoundSampler)
+	}
+	return false
+}
+
+// -----------------------
+
+type adaptiveSampler struct {
+	samplers map[string]Sampler
+
+	// TODO expand to keep track of newly seen operations
+	// TODO add a cap on the number of operations
+	// TODO need default rate limit and probabilistic rates for first time operations
+}
+
+// NewAdaptiveSampler adaptiveSampler is a delegating sampler that applies both probabilisticSampler and
+// rateLimitingSampler via the operationSampler. This sampler keeps track of all operations and delegates
+// calls to the respective operationSampler.
+func NewAdaptiveSampler(maxTracesPerSecond float64, samplingRates map[string]float64) (Sampler, error) {
+	samplers := make(map[string]Sampler)
+	for operation, samplingRate := range samplingRates {
+		probabilisticSampler, err := NewProbabilisticSampler(samplingRate)
+		if err != nil {
+			return nil, err
+		}
+		lowerBoundSampler, err := NewLowerBoundSampler(maxTracesPerSecond, samplingRate)
+		if err != nil {
+			return nil, err
+		}
+		samplers[operation] = &operationSampler{
+			probabilisticSampler: probabilisticSampler,
+			lowerBoundSampler:    lowerBoundSampler,
+			operation:            operation,
+		}
+	}
+	return &adaptiveSampler{
+		samplers: samplers,
+	}, nil
+}
+
+func (s *adaptiveSampler) IsSampled(id uint64, operation string) (bool, []Tag) {
+	sampler, ok := s.samplers[operation]
```

Comment:
not thread safe, must use rw locks. I've seen m3 guys do a double-checked locking, i.e.

* take read lock
* look for sampler in the map
* if not found, take write lock
* look for sampler again - if found, then return
* create new one and store in the map

above should be in a helper function.